session/filetransfer.go

```diff
@@ -0,0 +1,549 @@
+		}
+		if err != nil && err != errFileTransferAbort {
+			h.Error(msg, w, err)
+			log.Error(err.Error())
+		}
+	}()
+	select {
+	case h.mutex <- struct{}{}:
+		defer func() { <-h.mutex }()
+	default:
+		return errors.New("another file transfer is still in progress")
+	}
+
+	chunker := &chunkWriter{
+		SessionID: msg.Header.SessionID,
+		W:         w,
+	}
+
+	waitAck := func() (*ws.ProtoMsg, error) {
+		msg, open := <-h.msgChan
+		if !open {
+			return nil, errFileTransferAbort
+		}
+		switch msg.Header.MsgType {
+		case wsft.MessageTypeACK:
+
+		case wsft.MessageTypeError:
+			var erro wsft.Error
+			msgpack.Unmarshal(msg.Body, erro) //nolint:errcheck
+			if erro.Error != nil {
+				log.Errorf("received error message from client: %s", erro.Error)
+			} else {
+				log.Error("received malformed error message from client: aborting")
+			}
+			return msg, errFileTransferAbort
+
+		default:
+			return msg, errors.Errorf(
+				"received unexpected message type '%s' during download",
+			)
+		}
+		if off, ok := msg.Header.Properties["offset"].(int64); ok {
+			ackOffset = off
+		} else {
+			ackOffset += FileTransferBufSize
+		}
+		return msg, nil
+	}
+
+	buf := make([]byte, FileTransferBufSize)
+	for {
+		windowBytes := ackOffset - chunker.Offset +
+			ACKSlidingWindowRecv*FileTransferBufSize
+		if windowBytes > 0 {
+			N, err := io.CopyBuffer(chunker, io.LimitReader(fd, windowBytes), buf)
+			if err != nil {
+				err = errors.Wrap(err, "failed to copy file chunk to session")
+				return err
+			}
+			if N < windowBytes {
+				break
+			}
+		}
+
+		msg, err = waitAck()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Send EOF chunk
+	err = w.WriteProtoMsg(&ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeFileTransfer,
+			MsgType:   wsft.MessageTypeChunk,
+			SessionID: msg.Header.SessionID,
+			Properties: map[string]interface{}{
+				"offset": chunker.Offset,
+			},
+		},
+	})
+	if err != nil {
+		log.Errorf("failed to send EOF message to client: %s", err.Error())
+		return err
+	}
+
+	for ackOffset < chunker.Offset {
+		msg, err = waitAck()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *FileTransferHandler) InitFileUpload(msg *ws.ProtoMsg, w ResponseWriter) (err error) {
+	var (
+		defaultMode uint32 = 0644
+		defaultUID  uint32 = uint32(os.Getuid())
+		defaultGID  uint32 = uint32(os.Getgid())
+	)
+	params := FileInfo{
+		UID:  &defaultUID,
+		GID:  &defaultGID,
+		Mode: &defaultMode,
+	}
+	defer func() {
+		if err != nil {
+			h.Error(msg, w, err)
+		}
+	}()
+	err = msgpack.Unmarshal(msg.Body, &params)
+	if err != nil {
+		return errors.Wrap(err, "malformed request body")
+	} else if err = params.Validate(); err != nil {
+		return errors.Wrap(err, "invalid request parameters")
+	}
+
+	go h.FileUploadHandler(msg, params, w) //nolint:errcheck
+	return nil
+}
+
+func prepareTempFile(params FileInfo) (*os.File, error) {
+	fd, err := ioutil.TempFile(path.Dir(*params.Path), path.Base(*params.Path))
```

Comment:
two questions here:
1. is the file removed in all scenarios before we reach the final move to the target path?
2. this is not safe, since TempFile creation and the Chown below do not make an atomic operation; suggested change:
* set umask to `0077`
* create a temporary directory
* create all the temporary files in that directory
* once done make the final move, checking if the destination is not something suspicious
* remove the temporary directory recursively
* restore the umask
* set the modes and permissions on the destination file

Reply:
1. Good catch, this is indeed not the case if this function returns an error.
2. I'll try to avoid using `ioutil.Tempfile`, but I don't think it's a good idea to create the file outside the target directory (since we don't know if the target is on a different filesystem; nor if we have permission to create the file there at all).

Reply:
The solution I went for is creating a new temporary file in the target directory with `0200` permissions. Then, before I commit the file to the target path, I change the permissions and owner (in that order) to the ones provided in the initial request.